firmware/common/portapack_io.hpp

```diff
@@ -343,9 +343,15 @@ class IO {
         uint16_t g = (pixel.v >> 5) & 0x3F;   // Extract green
         uint16_t b = pixel.v & 0x1F;          // Extract blue
 
-        r = r >> darken_level_shift;  // Darken red
-        g = g >> darken_level_shift;  // Darken green
-        b = b >> darken_level_shift;  // Darken blue
+        if (shift_left) {  // Shfting
+            r = r << shift_level;
+            g = g << shift_level;
```

Comment:
I haven't tested this, but I was thinking that something like this might be slightly faster:

```
void shift_color(ui::Color& pixel, uint8_t shift_level, bool shift_left) {
    static uint16_t darken_mask[] = { 0b1111111111111111, 0b0111101111101111, 0b0011100111100111, 0b0001100011100011 };
    if (shift_left)
        pixel.v <<= shift_level;  // assuming here that we're un-darkening previously darkened pixels
    else
        pixel.v = (pixel.v >> shift_level) & darken_mask[shift_level];
}
```

Reply:
I’m sorry I didn’t understand this one yet. However since this one if faster, maybe you can open a new PR?

Reply:
I just tested on my device, it's really faster!!!!!! thank you sooooo much!
Since you wrote this code and i can't take the credit from you, i'm thinking that we have two ways to use it:
1. I close this PR and you do same change + your change in a new PR
2. We merge this one and you submit a new PR do your change

both ways are ok with me. how do you think?

Reply:
If my suggestion works, it would be more convenient if you went ahead and added it to this PR so I didn't have to submit one.  To save a few more instructions, you might also consider breaking it apart into 2 functions (eliminating the "if" statement and passing the "shift left" flag), or using a macro:

```
static const uint16_t darken_mask[] = { 0b1111111111111111, 0b0111101111101111, 0b0011100111100111, 0b0001100011100011 };

void darken(ui::Color& pixel, uint8_t shift_level) {
    pixel.v = (pixel.v >> shift_level) & darken_mask[shift_level];
}

void undarken(ui::Color& pixel, uint8_t shift_level) {
    pixel.v <<= shift_level;  // assuming here that we're un-darkening previously darkened pixels
}
```

Reply:
Your suggestion works really good and it’s indeed faster.💕

Considering the code credit, to me, it seems would be better you do a PR. 
Another reason is that I didn’t yet figured out why it’s faster, so I’m not sure if I can take it for granted. 
However if you would like to insist ask me to do it, then sure. I can still attach your username on comments (or if you prefer, I also can attach nothing, it’s your decision)

about the apart suggestion, I’m not sure why take them apart and use a new var would save some instruction (I barely don’t know compiler and assembly sorry😂) but I will take this advice for sure (if you still need me to add them)

thx!

Reply:
I would prefer if you just add my small suggestions in this PR.  If you don't want to do that, you can merge yours now and I'll submit another PR sometime later today.

Reply:
I will choose the first option, since it’s your code and I will felt guilty if I took it.